File: gpurec/core/batch_planning.py
"""Shared family batch planning for resident and chunked model APIs.

This module is a narrow low-level support boundary for in-repo API, workflow,
CLI, memory-policy, and white-box test callers.  The exported names are shared
planning helpers, not a promise that the rest of ``gpurec.core`` is stable.
"""
from __future__ import annotations

from dataclasses import dataclass
import math
from numbers import Integral, Real
import os
from typing import Sequence


_SCHEDULER_BACKEND_ENV = "GPUREC_SCHEDULER_BACKEND"


@dataclass(frozen=True)
class FamilyBatchPlan:
    """One planned family batch before tensors and wave layouts are built."""

    indices: list[int]
    clades: int
    splits: int


def normalize_batch_packing(value: str | None) -> str:
    """Normalize user-facing batch-packing aliases."""
    if value is None:
        return "sequential"
    text = str(value).strip().lower().replace("-", "_")
    aliases = {
        "": "sequential",
        "sequential": "sequential",
        "contiguous": "sequential",
        "input_order": "sequential",
        "clade_first_fit": "clade_first_fit",
        "first_fit_decreasing": "clade_first_fit",
        "ffd": "clade_first_fit",
        "clade_ffd": "clade_first_fit",
        "depth_first_fit": "depth_first_fit",
        "depth_ffd": "depth_first_fit",
        "critical_path_first_fit": "depth_first_fit",
        "critical_first_fit": "depth_first_fit",
        "wave_first_fit": "depth_first_fit",
    }
    try:
        return aliases[text]
    except KeyError as exc:
        raise ValueError(
            "batch_packing must be 'sequential', 'clade_first_fit', or "
            f"'depth_first_fit', got {value!r}"
        ) from exc


def _normalize_int_control(name: str, value: int | float | str) -> int:
    if isinstance(value, bool):
        raise ValueError(f"{name} must be an integer")
    if isinstance(value, str):
        try:
            return int(value.strip())
        except ValueError as exc:
            raise ValueError(f"{name} must be an integer") from exc
    if isinstance(value, Integral):
        return int(value)
    if isinstance(value, Real):
        number = float(value)
        if not math.isfinite(number) or not number.is_integer():
            raise ValueError(f"{name} must be an integer")
        return int(number)
    raise ValueError(f"{name} must be an integer")


def normalize_clade_budget(value: int | float | str | None) -> int | None:
    """Normalize an optional positive clade budget."""
    if value is None:
        return None
    budget = _normalize_int_control("clade_budget", value)
    if budget <= 0:
        raise ValueError("clade_budget must be positive when provided")
    return budget


def normalize_family_chunk_size(
    value: int | float | str | None,
    *,
    allow_auto: bool = False,
) -> int | str:
    """Normalize family chunk-size controls shared by APIs and workflows."""
    if value is None:
        return 0
    if isinstance(value, str):
        text = value.strip().lower()
        if text in {"", "0", "all", "none", "null"}:
            return 0
        if text == "auto":
            if allow_auto:
                return "auto"
            raise ValueError(
                "family chunk size 'auto' is not supported; use 0 for one "
                "resident batch or a positive integer"
            )
    size = _normalize_int_control("family_chunk_size", value)
    if size < 0:
        raise ValueError("family_chunk_size must be non-negative")
    return size


def _selected_indices(
    *,
    indices: Sequence[int] | None,
    total: int | None,
    clade_counts: Sequence[int],
) -> list[int]:
    if indices is not None:
        return [
            _normalize_int_control("family index", idx)
            for idx in indices
        ]
    if total is None:
        total = len(clade_counts)
    total_int = _normalize_int_control("total", total)
    if total_int < 0:
        raise ValueError("total must be non-negative")
    return list(range(total_int))


def _validate_selected_indices(selected: Sequence[int], clade_counts: Sequence[int]) -> None:
    limit = len(clade_counts)
    seen: set[int] = set()
    for position, idx in enumerate(selected):
        if idx < 0 or idx >= limit:
            raise ValueError(
                f"family index {idx} at selected position {position} is outside "
                f"valid range [0, {limit})"
            )
        if idx in seen:
            raise ValueError(f"duplicate family index {idx} at selected position {position}")
        seen.add(idx)


def _require_indexed_stats(
    name: str,
    values: Sequence[int] | None,
    selected: Sequence[int],
) -> Sequence[int]:
    if values is None:
        raise ValueError(f"batch_packing='depth_first_fit' requires {name}")
    if selected:
        required = max(selected) + 1
        if len(values) < required:
            raise ValueError(f"{name} must cover selected family indices")
    return values


def _plan_from_chunks(
    chunks: Sequence[Sequence[int]],
    clade_counts: Sequence[int],
    split_counts: Sequence[int] | None,
) -> list[FamilyBatchPlan]:
    plans: list[FamilyBatchPlan] = []
    for chunk in chunks:
        indices = [int(idx) for idx in chunk]
        plans.append(
            FamilyBatchPlan(
                indices=indices,
                clades=sum(int(clade_counts[idx]) for idx in indices),
                splits=(
                    0
                    if split_counts is None
                    else sum(int(split_counts[idx]) for idx in indices)
                ),
            )
        )
    return plans


def _plan_family_batches_python(
    *,
    clade_counts: Sequence[int],
    family_chunk_size: int,
    clade_budget: int | None,
    batch_packing: str = "sequential",
    indices: Sequence[int] | None = None,
    total: int | None = None,
    split_counts: Sequence[int] | None = None,
    leaf_counts: Sequence[int] | None = None,
    nonleaf_counts: Sequence[int] | None = None,
    schedule_depths: Sequence[int] | None = None,
    max_wave_size: int | None = None,
) -> list[FamilyBatchPlan]:
    """Plan family batches using a shared packing implementation.

    ``indices`` may select a subset of families, but all count arrays are
    indexed by the original family index.
    """
    selected = _selected_indices(
        indices=indices,
        total=total,
        clade_counts=clade_counts,
    )
    _validate_selected_indices(selected, clade_counts)
    family_limit = _normalize_int_control("family_chunk_size", family_chunk_size)
    if family_limit < 0:
        raise ValueError("family_chunk_size must be non-negative")

    packing = normalize_batch_packing(batch_packing)
    budget = normalize_clade_budget(clade_budget)

    if packing == "clade_first_fit":
        if budget is None:
            raise ValueError("batch_packing='clade_first_fit' requires clade_budget")
        chunks: list[list[int]] = []
        chunk_clades: list[int] = []
        order = sorted(selected, key=lambda idx: int(clade_counts[idx]), reverse=True)
        for idx in order:
            n_clades = int(clade_counts[idx])
            best_j: int | None = None
            best_remaining: int | None = None
            for j, current_clades in enumerate(chunk_clades):
                if family_limit > 0 and len(chunks[j]) >= family_limit:
                    continue
                remaining = budget - current_clades - n_clades
                if remaining < 0:
                    continue
                if best_remaining is None or remaining < best_remaining:
                    best_j = j
                    best_remaining = remaining
            if best_j is None:
                chunks.append([int(idx)])
                chunk_clades.append(n_clades)
            else:
                chunks[best_j].append(int(idx))
                chunk_clades[best_j] += n_clades
        return _plan_from_chunks(chunks, clade_counts, split_counts)

    if packing == "depth_first_fit":
        if budget is None:
            raise ValueError("batch_packing='depth_first_fit' requires clade_budget")
        leaves = _require_indexed_stats("leaf_counts", leaf_counts, selected)
        nonleaves = _require_indexed_stats("nonleaf_counts", nonleaf_counts, selected)
        depths = _require_indexed_stats("schedule_depths", schedule_depths, selected)

        wave_cap = (
            sum(int(clade_counts[idx]) for idx in selected)
            if max_wave_size is None
            else _normalize_int_control("max_wave_size", max_wave_size)
        )
        if wave_cap <= 0:
            raise ValueError("max_wave_size must be positive")

        def lower_bound(leaves_count: int, nonleaves_count: int, depth: int) -> int:
            leaf_waves = math.ceil(int(leaves_count) / wave_cap)
            work_waves = math.ceil(int(nonleaves_count) / wave_cap)
            return leaf_waves + max(int(depth), work_waves)

        chunks: list[list[int]] = []
        chunk_clades: list[int] = []
        chunk_leaves: list[int] = []
        chunk_nonleaves: list[int] = []
        chunk_depths: list[int] = []
        order = sorted(
            selected,
            key=lambda idx: (int(depths[idx]), int(clade_counts[idx])),
            reverse=True,
        )
        for idx in order:
            n_clades = int(clade_counts[idx])
            n_leaves = int(leaves[idx])
            n_nonleaves = int(nonleaves[idx])
            depth = int(depths[idx])
            best_j: int | None = None
            best_key: tuple[int, int, int] | None = None
            for j, current_clades in enumerate(chunk_clades):
                if family_limit > 0 and len(chunks[j]) >= family_limit:
                    continue
                new_clades = current_clades + n_clades
                if new_clades > budget:
                    continue
                before = lower_bound(
                    chunk_leaves[j],
                    chunk_nonleaves[j],
                    chunk_depths[j],
                )
                after = lower_bound(
                    chunk_leaves[j] + n_leaves,
                    chunk_nonleaves[j] + n_nonleaves,
                    max(chunk_depths[j], depth),
                )
                remaining = budget - new_clades
                key = (after - before, after, remaining)
                if best_key is None or key < best_key:
                    best_j = j
                    best_key = key
            if best_j is None:
                chunks.append([int(idx)])
                chunk_clades.append(n_clades)
                chunk_leaves.append(n_leaves)
                chunk_nonleaves.append(n_nonleaves)
                chunk_depths.append(depth)
            else:
                chunks[best_j].append(int(idx))
                chunk_clades[best_j] += n_clades
                chunk_leaves[best_j] += n_leaves
                chunk_nonleaves[best_j] += n_nonleaves
                chunk_depths[best_j] = max(chunk_depths[best_j], depth)
        return _plan_from_chunks(chunks, clade_counts, split_counts)

    chunks: list[list[int]] = []
    current: list[int] = []
    current_clades = 0

    def flush() -> None:
        nonlocal current, current_clades
        if current:
            chunks.append(list(current))
            current = []
            current_clades = 0

    for idx in selected:
        n_clades = int(clade_counts[idx])
        family_cap_hit = family_limit > 0 and len(current) >= family_limit
        clade_cap_hit = (
            budget is not None
            and current
            and current_clades + n_clades > budget
        )
        if family_cap_hit or clade_cap_hit:
            flush()
        current.append(int(idx))
        current_clades += n_clades
    flush()
    return _plan_from_chunks(chunks, clade_counts, split_counts)


def plan_family_batches(
    *,
    clade_counts: Sequence[int],
    family_chunk_size: int,
    clade_budget: int | None,
    batch_packing: str = "sequential",
    indices: Sequence[int] | None = None,
    total: int | None = None,
    split_counts: Sequence[int] | None = None,
    leaf_counts: Sequence[int] | None = None,
    nonleaf_counts: Sequence[int] | None = None,
    schedule_depths: Sequence[int] | None = None,
    max_wave_size: int | None = None,
) -> list[FamilyBatchPlan]:
    """Plan family batches using the configured scheduling backend."""
    backend = os.environ.get(_SCHEDULER_BACKEND_ENV, "rust").strip().lower()
    if backend in {"python", "py"}:
        return _plan_family_batches_python(
            clade_counts=clade_counts,
            family_chunk_size=family_chunk_size,
            clade_budget=clade_budget,
            batch_packing=batch_packing,
            indices=indices,
            total=total,
            split_counts=split_counts,
            leaf_counts=leaf_counts,
            nonleaf_counts=nonleaf_counts,
            schedule_depths=schedule_depths,
            max_wave_size=max_wave_size,
        )
    if backend in {"", "rust"}:
        from gpurec.core.schedule_rust import plan_family_batches as rust_plan

        return [
            FamilyBatchPlan(
                indices=list(plan["indices"]),
                clades=int(plan["clades"]),
                splits=int(plan["splits"]),
            )
            for plan in rust_plan(
                clade_counts=clade_counts,
                family_chunk_size=family_chunk_size,
                clade_budget=clade_budget,
                batch_packing=batch_packing,
                indices=indices,
                total=total,
                split_counts=split_counts,
                leaf_counts=leaf_counts,
                nonleaf_counts=nonleaf_counts,
                schedule_depths=schedule_depths,
                max_wave_size=max_wave_size,
            )
        ]
    raise ValueError(
        f"{_SCHEDULER_BACKEND_ENV} must be 'python' or 'rust', got {backend!r}"
    )


__all__ = [
    "FamilyBatchPlan",
    "normalize_batch_packing",
    "normalize_clade_budget",
    "normalize_family_chunk_size",
    "plan_family_batches",
]

File: crates/gpurec-preprocess/src/batch_planning.rs
use crate::PreprocessError;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize)]
pub struct BatchPlanRequest {
    pub clade_counts: Vec<i64>,
    pub family_chunk_size: i64,
    #[serde(default)]
    pub clade_budget: Option<i64>,
    #[serde(default = "default_batch_packing")]
    pub batch_packing: String,
    #[serde(default)]
    pub indices: Option<Vec<i64>>,
    #[serde(default)]
    pub total: Option<i64>,
    #[serde(default)]
    pub split_counts: Option<Vec<i64>>,
    #[serde(default)]
    pub leaf_counts: Option<Vec<i64>>,
    #[serde(default)]
    pub small_family_max_leaves: Option<i64>,
    #[serde(default)]
    pub nonleaf_counts: Option<Vec<i64>>,
    #[serde(default)]
    pub schedule_depths: Option<Vec<i64>>,
    #[serde(default)]
    pub max_wave_size: Option<i64>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct FamilyBatchPlanOutput {
    pub indices: Vec<i64>,
    pub clades: i64,
    pub splits: i64,
}

fn default_batch_packing() -> String {
    "sequential".to_string()
}

pub fn plan_family_batches_request(
    request: &BatchPlanRequest,
) -> Result<Vec<FamilyBatchPlanOutput>, PreprocessError> {
    plan_family_batches_impl(
        &request.clade_counts,
        request.family_chunk_size,
        request.clade_budget,
        &request.batch_packing,
        request.indices.as_deref(),
        request.total,
        request.split_counts.as_deref(),
        request.leaf_counts.as_deref(),
        request.nonleaf_counts.as_deref(),
        request.schedule_depths.as_deref(),
        request.max_wave_size,
        request.small_family_max_leaves,
    )
}

#[allow(clippy::too_many_arguments)]
pub fn plan_family_batches(
    clade_counts: &[i64],
    family_chunk_size: i64,
    clade_budget: Option<i64>,
    batch_packing: &str,
    indices: Option<&[i64]>,
    total: Option<i64>,
    split_counts: Option<&[i64]>,
    leaf_counts: Option<&[i64]>,
    nonleaf_counts: Option<&[i64]>,
    schedule_depths: Option<&[i64]>,
    max_wave_size: Option<i64>,
) -> Result<Vec<FamilyBatchPlanOutput>, PreprocessError> {
    plan_family_batches_impl(
        clade_counts,
        family_chunk_size,
        clade_budget,
        batch_packing,
        indices,
        total,
        split_counts,
        leaf_counts,
        nonleaf_counts,
        schedule_depths,
        max_wave_size,
        None,
    )
}

#[allow(clippy::too_many_arguments)]
fn plan_family_batches_impl(
    clade_counts: &[i64],
    family_chunk_size: i64,
    clade_budget: Option<i64>,
    batch_packing: &str,
    indices: Option<&[i64]>,
    total: Option<i64>,
    split_counts: Option<&[i64]>,
    leaf_counts: Option<&[i64]>,
    nonleaf_counts: Option<&[i64]>,
    schedule_depths: Option<&[i64]>,
    max_wave_size: Option<i64>,
    small_family_max_leaves: Option<i64>,
) -> Result<Vec<FamilyBatchPlanOutput>, PreprocessError> {
    let selected = selected_indices(indices, total, clade_counts.len())?;
    validate_selected_indices(&selected, clade_counts.len())?;
    if let Some(counts) = split_counts {
        require_indexed_stats("split_counts", "split_counts", Some(counts), &selected)?;
    }
    if family_chunk_size < 0 {
        return invalid("family_chunk_size must be non-negative");
    }
    let packing = normalize_batch_packing(batch_packing)?;
    if let Some(budget) = clade_budget {
        if budget <= 0 {
            return invalid("clade_budget must be positive when provided");
        }
    }
    if let Some(max_leaves) = small_family_max_leaves {
        if max_leaves < 0 {
            return invalid("small_family_max_leaves must be non-negative when provided");
        }
    }

    if packing == "clade_first_fit" {
        clade_budget.ok_or_else(|| {
            PreprocessError::InvalidInput(
                "batch_packing='clade_first_fit' requires clade_budget".to_string(),
            )
        })?;
    }
    if packing == "depth_first_fit" {
        clade_budget.ok_or_else(|| {
            PreprocessError::InvalidInput(
                "batch_packing='depth_first_fit' requires clade_budget".to_string(),
            )
        })?;
        require_indexed_stats(
            "batch_packing='depth_first_fit'",
            "leaf_counts",
            leaf_counts,
            &selected,
        )?;
        require_indexed_stats(
            "batch_packing='depth_first_fit'",
            "nonleaf_counts",
            nonleaf_counts,
            &selected,
        )?;
        require_indexed_stats(
            "batch_packing='depth_first_fit'",
            "schedule_depths",
            schedule_depths,
            &selected,
        )?;
    }

    let resolved_max_wave_size = if packing == "depth_first_fit" {
        let wave_cap = match max_wave_size {
            Some(value) => value,
            None => selected.iter().map(|idx| clade_counts[*idx]).sum(),
        };
        if wave_cap <= 0 {
            return invalid("max_wave_size must be positive");
        }
        Some(wave_cap)
    } else {
        max_wave_size
    };

    let small_family_leaf_counts = match small_family_max_leaves {
        Some(_) => Some(require_indexed_stats(
            "small_family_max_leaves",
            "leaf_counts",
            leaf_counts,
            &selected,
        )?),
        None => None,
    };
    let selected_groups =
        selected_groups_by_leaf_count(&selected, small_family_leaf_counts, small_family_max_leaves);

    let mut plans = Vec::new();
    for group in selected_groups {
        plans.extend(plan_family_batches_for_selected(
            &group,
            clade_counts,
            split_counts,
            family_chunk_size,
            clade_budget,
            packing,
            leaf_counts,
            nonleaf_counts,
            schedule_depths,
            resolved_max_wave_size,
        )?);
    }
    Ok(plans)
}

#[allow(clippy::too_many_arguments)]
fn plan_family_batches_for_selected(
    selected: &[usize],
    clade_counts: &[i64],
    split_counts: Option<&[i64]>,
    family_chunk_size: i64,
    clade_budget: Option<i64>,
    packing: &str,
    leaf_counts: Option<&[i64]>,
    nonleaf_counts: Option<&[i64]>,
    schedule_depths: Option<&[i64]>,
    max_wave_size: Option<i64>,
) -> Result<Vec<FamilyBatchPlanOutput>, PreprocessError> {
    match packing {
        "clade_first_fit" => {
            let budget = clade_budget.ok_or_else(|| {
                PreprocessError::InvalidInput(
                    "batch_packing='clade_first_fit' requires clade_budget".to_string(),
                )
            })?;
            plan_clade_first_fit(
                selected,
                clade_counts,
                split_counts,
                family_chunk_size,
                budget,
            )
        }
        "depth_first_fit" => {
            let budget = clade_budget.ok_or_else(|| {
                PreprocessError::InvalidInput(
                    "batch_packing='depth_first_fit' requires clade_budget".to_string(),
                )
            })?;
            let leaves = require_indexed_stats(
                "batch_packing='depth_first_fit'",
                "leaf_counts",
                leaf_counts,
                selected,
            )?;
            let nonleaves = require_indexed_stats(
                "batch_packing='depth_first_fit'",
                "nonleaf_counts",
                nonleaf_counts,
                selected,
            )?;
            let depths = require_indexed_stats(
                "batch_packing='depth_first_fit'",
                "schedule_depths",
                schedule_depths,
                selected,
            )?;
            let wave_cap = max_wave_size.ok_or_else(|| {
                PreprocessError::InvalidInput(
                    "batch_packing='depth_first_fit' requires max_wave_size".to_string(),
                )
            })?;
            plan_depth_first_fit(
                selected,
                clade_counts,
                split_counts,
                leaves,
                nonleaves,
                depths,
                family_chunk_size,
                budget,
                wave_cap,
            )
        }
        "sequential" => plan_sequential(
            selected,
            clade_counts,
            split_counts,
            family_chunk_size,
            clade_budget,
        ),
        _ => invalid(format!("unknown batch_packing {packing:?}")),
    }
}

fn selected_groups_by_leaf_count(
    selected: &[usize],
    leaf_counts: Option<&[i64]>,
    small_family_max_leaves: Option<i64>,
) -> Vec<Vec<usize>> {
    let Some(max_leaves) = small_family_max_leaves else {
        return vec![selected.to_vec()];
    };
    if max_leaves == 0 {
        return vec![selected.to_vec()];
    }
    let Some(leaves) = leaf_counts else {
        return vec![selected.to_vec()];
    };
    let mut boundaries = vec![max_leaves];
    let mut next = max_leaves;
    while next < 256 {
        next *= 2;
        boundaries.push(next);
    }
    boundaries.push(i64::MAX);
    let mut groups = vec![Vec::new(); boundaries.len()];
    for idx in selected {
        let leaf_count = leaves[*idx];
        let bucket = boundaries
            .iter()
            .position(|boundary| leaf_count <= *boundary)
            .unwrap_or(boundaries.len() - 1);
        groups[bucket].push(*idx);
    }
    groups
        .into_iter()
        .filter(|group| !group.is_empty())
        .collect()
}

fn invalid<T>(message: impl Into<String>) -> Result<T, PreprocessError> {
    Err(PreprocessError::InvalidInput(message.into()))
}

fn normalize_batch_packing(value: &str) -> Result<&'static str, PreprocessError> {
    let text = value.trim().to_ascii_lowercase().replace('-', "_");
    match text.as_str() {
        "" | "sequential" | "contiguous" | "input_order" => Ok("sequential"),
        "clade_first_fit" | "first_fit_decreasing" | "ffd" | "clade_ffd" => Ok("clade_first_fit"),
        "depth_first_fit"
        | "depth_ffd"
        | "critical_path_first_fit"
        | "critical_first_fit"
        | "wave_first_fit" => Ok("depth_first_fit"),
        _ => invalid(format!(
            "batch_packing must be 'sequential', 'clade_first_fit', or \
             'depth_first_fit', got {value:?}"
        )),
    }
}

fn selected_indices(
    indices: Option<&[i64]>,
    total: Option<i64>,
    clade_count_len: usize,
) -> Result<Vec<usize>, PreprocessError> {
    if let Some(indices) = indices {
        return indices
            .iter()
            .map(|idx| {
                if *idx < 0 {
                    invalid(format!("family index {idx} is outside valid range"))
                } else {
                    Ok(*idx as usize)
                }
            })
            .collect();
    }
    let total = total.unwrap_or(clade_count_len as i64);
    if total < 0 {
        return invalid("total must be non-negative");
    }
    Ok((0..total as usize).collect())
}

fn validate_selected_indices(
    selected: &[usize],
    clade_count_len: usize,
) -> Result<(), PreprocessError> {
    let mut seen = std::collections::HashSet::with_capacity(selected.len());
    for (position, idx) in selected.iter().enumerate() {
        if *idx >= clade_count_len {
            return invalid(format!(
                "family index {idx} at selected position {position} is outside valid range [0, {clade_count_len})"
            ));
        }
        if !seen.insert(*idx) {
            return invalid(format!(
                "duplicate family index {idx} at selected position {position}"
            ));
        }
    }
    Ok(())
}

fn require_indexed_stats<'a>(
    context: &str,
    name: &str,
    values: Option<&'a [i64]>,
    selected: &[usize],
) -> Result<&'a [i64], PreprocessError> {
    let values = values
        .ok_or_else(|| PreprocessError::InvalidInput(format!("{context} requires {name}")))?;
    if let Some(required) = selected.iter().max().map(|idx| idx + 1) {
        if values.len() < required {
            return invalid(format!("{name} must cover selected family indices"));
        }
    }
    Ok(values)
}

fn plan_from_chunks(
    chunks: Vec<Vec<usize>>,
    clade_counts: &[i64],
    split_counts: Option<&[i64]>,
) -> Vec<FamilyBatchPlanOutput> {
    chunks
        .into_iter()
        .map(|chunk| {
            let clades = chunk.iter().map(|idx| clade_counts[*idx]).sum();
            let splits = split_counts
                .map(|counts| chunk.iter().map(|idx| counts[*idx]).sum())
                .unwrap_or(0);
            FamilyBatchPlanOutput {
                indices: chunk.into_iter().map(|idx| idx as i64).collect(),
                clades,
                splits,
            }
        })
        .collect()
}

fn plan_clade_first_fit(
    selected: &[usize],
    clade_counts: &[i64],
    split_counts: Option<&[i64]>,
    family_limit: i64,
    budget: i64,
) -> Result<Vec<FamilyBatchPlanOutput>, PreprocessError> {
    let mut chunks: Vec<Vec<usize>> = Vec::new();
    let mut chunk_clades: Vec<i64> = Vec::new();
    let mut order = selected.to_vec();
    order.sort_by(|left, right| clade_counts[*right].cmp(&clade_counts[*left]));

    for idx in order {
        let n_clades = clade_counts[idx];
        let mut best_j = None;
        let mut best_remaining = None;
        for (j, current_clades) in chunk_clades.iter().enumerate() {
            if family_limit > 0 && chunks[j].len() >= family_limit as usize {
                continue;
            }
            let remaining = budget - current_clades - n_clades;
            if remaining < 0 {
                continue;
            }
            if best_remaining.map(|best| remaining < best).unwrap_or(true) {
                best_j = Some(j);
                best_remaining = Some(remaining);
            }
        }
        if let Some(j) = best_j {
            chunks[j].push(idx);
            chunk_clades[j] += n_clades;
        } else {
            chunks.push(vec![idx]);
            chunk_clades.push(n_clades);
        }
    }
    Ok(plan_from_chunks(chunks, clade_counts, split_counts))
}

#[allow(clippy::too_many_arguments)]
fn plan_depth_first_fit(
    selected: &[usize],
    clade_counts: &[i64],
    split_counts: Option<&[i64]>,
    leaf_counts: &[i64],
    nonleaf_counts: &[i64],
    schedule_depths: &[i64],
    family_limit: i64,
    budget: i64,
    wave_cap: i64,
) -> Result<Vec<FamilyBatchPlanOutput>, PreprocessError> {
    let mut chunks: Vec<Vec<usize>> = Vec::new();
    let mut chunk_clades: Vec<i64> = Vec::new();
    let mut chunk_leaves: Vec<i64> = Vec::new();
    let mut chunk_nonleaves: Vec<i64> = Vec::new();
    let mut chunk_depths: Vec<i64> = Vec::new();
    let mut order = selected.to_vec();
    order.sort_by(|left, right| {
        (schedule_depths[*right], clade_counts[*right])
            .cmp(&(schedule_depths[*left], clade_counts[*left]))
    });

    for idx in order {
        let n_clades = clade_counts[idx];
        let n_leaves = leaf_counts[idx];
        let n_nonleaves = nonleaf_counts[idx];
        let depth = schedule_depths[idx];
        let mut best_j = None;
        let mut best_key: Option<(i64, i64, i64)> = None;
        for (j, current_clades) in chunk_clades.iter().enumerate() {
            if family_limit > 0 && chunks[j].len() >= family_limit as usize {
                continue;
            }
            let new_clades = current_clades + n_clades;
            if new_clades > budget {
                continue;
            }
            let before = lower_bound(
                chunk_leaves[j],
                chunk_nonleaves[j],
                chunk_depths[j],
                wave_cap,
            );
            let after = lower_bound(
                chunk_leaves[j] + n_leaves,
                chunk_nonleaves[j] + n_nonleaves,
                chunk_depths[j].max(depth),
                wave_cap,
            );
            let remaining = budget - new_clades;
            let key = (after - before, after, remaining);
            if best_key.map(|best| key < best).unwrap_or(true) {
                best_j = Some(j);
                best_key = Some(key);
            }
        }
        if let Some(j) = best_j {
            chunks[j].push(idx);
            chunk_clades[j] += n_clades;
            chunk_leaves[j] += n_leaves;
            chunk_nonleaves[j] += n_nonleaves;
            chunk_depths[j] = chunk_depths[j].max(depth);
        } else {
            chunks.push(vec![idx]);
            chunk_clades.push(n_clades);
            chunk_leaves.push(n_leaves);
            chunk_nonleaves.push(n_nonleaves);
            chunk_depths.push(depth);
        }
    }
    Ok(plan_from_chunks(chunks, clade_counts, split_counts))
}

fn lower_bound(leaves_count: i64, nonleaves_count: i64, depth: i64, wave_cap: i64) -> i64 {
    let leaf_waves = ceil_div_nonnegative(leaves_count, wave_cap);
    let work_waves = ceil_div_nonnegative(nonleaves_count, wave_cap);
    leaf_waves + depth.max(work_waves)
}

fn ceil_div_nonnegative(value: i64, divisor: i64) -> i64 {
    if value <= 0 {
        0
    } else {
        (value + divisor - 1) / divisor
    }
}

fn plan_sequential(
    selected: &[usize],
    clade_counts: &[i64],
    split_counts: Option<&[i64]>,
    family_limit: i64,
    budget: Option<i64>,
) -> Result<Vec<FamilyBatchPlanOutput>, PreprocessError> {
    let mut chunks: Vec<Vec<usize>> = Vec::new();
    let mut current: Vec<usize> = Vec::new();
    let mut current_clades = 0i64;

    for idx in selected {
        let n_clades = clade_counts[*idx];
        let family_cap_hit = family_limit > 0 && current.len() >= family_limit as usize;
        let clade_cap_hit = budget
            .map(|budget| !current.is_empty() && current_clades + n_clades > budget)
            .unwrap_or(false);
        if family_cap_hit || clade_cap_hit {
            chunks.push(std::mem::take(&mut current));
            current_clades = 0;
        }
        current.push(*idx);
        current_clades += n_clades;
    }
    if !current.is_empty() {
        chunks.push(current);
    }

    Ok(plan_from_chunks(chunks, clade_counts, split_counts))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn plan_indices(plans: &[FamilyBatchPlanOutput]) -> Vec<Vec<i64>> {
        plans.iter().map(|plan| plan.indices.clone()).collect()
    }

    #[test]
    fn clade_first_fit_matches_python_fixture() {
        let plans = plan_family_batches(
            &[8, 7, 6, 5, 4],
            0,
            Some(12),
            "clade_first_fit",
            None,
            Some(5),
            None,
            None,
            None,
            None,
            None,
        )
        .unwrap();
        assert_eq!(plan_indices(&plans), vec![vec![0, 4], vec![1, 3], vec![2]]);
    }

    #[test]
    fn depth_first_fit_matches_python_fixture() {
        let plans = plan_family_batches(
            &[6, 6, 6, 6, 6],
            0,
            Some(12),
            "depth_first_fit",
            None,
            Some(5),
            Some(&[10, 20, 30, 40, 50]),
            Some(&[1, 1, 1, 1, 1]),
            Some(&[5, 5, 5, 5, 5]),
            Some(&[10, 9, 2, 1, 1]),
            Some(8),
        )
        .unwrap();
        assert_eq!(plan_indices(&plans), vec![vec![0, 1], vec![2, 3], vec![4]]);
        assert_eq!(
            plans.iter().map(|plan| plan.splits).collect::<Vec<_>>(),
            vec![30, 70, 50]
        );
    }

    #[test]
    fn request_prioritizes_small_families_under_leaf_threshold() {
        let request = BatchPlanRequest {
            clade_counts: vec![5, 3, 4, 4, 6, 2],
            family_chunk_size: 2,
            clade_budget: Some(6),
            batch_packing: "clade_first_fit".to_string(),
            indices: None,
            total: Some(6),
            split_counts: None,
            leaf_counts: Some(vec![9, 4, 7, 2, 6, 4]),
            small_family_max_leaves: Some(4),
            nonleaf_counts: None,
            schedule_depths: None,
            max_wave_size: None,
        };

        let plans = plan_family_batches_request(&request).unwrap();

        assert_eq!(
            plan_indices(&plans),
            vec![vec![3, 5], vec![1], vec![4], vec![2], vec![0]]
        );
        assert!(plans.iter().all(|plan| plan.clades <= 6));
        assert!(plans.iter().all(|plan| plan.indices.len() <= 2));
    }

    #[test]
    fn request_buckets_families_by_leaf_count_bands() {
        let request = BatchPlanRequest {
            clade_counts: vec![2, 2, 2, 2, 2, 2],
            family_chunk_size: 0,
            clade_budget: Some(12),
            batch_packing: "clade_first_fit".to_string(),
            indices: None,
            total: Some(6),
            split_counts: None,
            leaf_counts: Some(vec![4, 5, 8, 9, 16, 33]),
            small_family_max_leaves: Some(4),
            nonleaf_counts: None,
            schedule_depths: None,
            max_wave_size: None,
        };

        let plans = plan_family_batches_request(&request).unwrap();

        assert_eq!(
            plan_indices(&plans),
            vec![vec![0], vec![1, 2], vec![3, 4], vec![5]]
        );
    }
}
